runtime/tests/runtime_integration_tests/test_reallocate_dpe_context_limits.rs

```diff
@@ -0,0 +1,296 @@
+    const BASE_DERIVE_CONTEXT_CMD: DeriveContextCmd = DeriveContextCmd {
+        handle: ContextHandle::default(),
+        data: [0; DPE_PROFILE.get_tci_size()],
+        flags: DeriveContextFlags::MAKE_DEFAULT,
+        tci_type: 0,
+        target_locality: 0,
+    };
+
+    // 32 contexts = 1 root node (PL0)+
+    //               1 rt_journey (PL0)
+    //               (pl0_limit - 2) PL0 contexts in loop +
+    //               1 PL1 context as transition +
+    //               (pl1_limit - 1) PL1 contexts in loop
+
+    // Fill PL0 contexts
+    for _ in 0..(pl0_limit - 2) {
+        let _ = execute_dpe_cmd(
+            model,
+            &mut Command::DeriveContext(&BASE_DERIVE_CONTEXT_CMD),
+            DpeResult::Success,
+        );
+    }
+
+    // Trigger failure by trying to derive one more context to PL0
+    let _ = execute_dpe_cmd(
+        model,
+        &mut Command::DeriveContext(&BASE_DERIVE_CONTEXT_CMD),
+        DpeResult::MboxCmdFailure(CaliptraError::RUNTIME_PL0_USED_DPE_CONTEXT_THRESHOLD_REACHED),
+    );
+
+    // Fill PL1 contexts
+    if pl1_limit > 0 {
+        for i in 0..pl1_limit {
+            if i == 0 {
+                // Transition to PL1 locality on first iteration
+                let derive_ctx_cmd = DeriveContextCmd {
+                    flags: DeriveContextFlags::MAKE_DEFAULT | DeriveContextFlags::CHANGE_LOCALITY,
+                    target_locality: 2,
+                    ..BASE_DERIVE_CONTEXT_CMD
+                };
+                let _ = execute_dpe_cmd(
+                    model,
+                    &mut Command::DeriveContext(&derive_ctx_cmd),
+                    DpeResult::Success,
+                );
+
+                model.set_apb_pauser(2);
+            } else {
+                let _ = execute_dpe_cmd(
+                    model,
+                    &mut Command::DeriveContext(&BASE_DERIVE_CONTEXT_CMD),
+                    DpeResult::Success,
+                );
+            }
+        }
+
+        // Trigger failure by trying to derive one more context to PL1
+        let _ = execute_dpe_cmd(
+            model,
+            &mut Command::DeriveContext(&BASE_DERIVE_CONTEXT_CMD),
+            DpeResult::MboxCmdFailure(
+                CaliptraError::RUNTIME_PL1_USED_DPE_CONTEXT_THRESHOLD_REACHED,
+            ),
+        );
+    }
+}
+
+fn reallocate_pl0_pl1_dpe_contexts(
+    model: &mut DefaultHwModel,
+    pl0_limit: u32,
+) -> Result<Option<Vec<u8>>, ModelError> {
+    let mut cmd = MailboxReq::ReallocateDpeContextLimits(ReallocateDpeContextLimitsReq {
+        hdr: MailboxReqHeader { chksum: 0 },
+        pl0_context_limit: pl0_limit,
+    });
+    cmd.populate_chksum().unwrap();
+
+    model.mailbox_execute(
+        u32::from(CommandId::REALLOCATE_DPE_CONTEXT_LIMITS),
+        cmd.as_bytes().unwrap(),
+    )
+}
+
+#[test]
+fn test_pl0_pl1_reallocation_range() {
+    for pl0_limit in 2..32 {
```

Comment:
You could use the DPE definition for the maximum number of contexts here. But on the other hand, we are planning to bump that value to 64 for 2.x and it would be nice to see that fail here.

```suggestion
    for pl0_limit in 2..MAX_HANDLES {
```